### PR TITLE
WPScan - implement confidence

### PR DESCRIPTION
**Pull request by damiencarol**
## Work done

- [x] Implement scanner confidence for WPScan and fix a bug in `models.py` when confidence is max
- [x] Implement `date` for findings
- [x] Add more real data to test for unit tests (empty file and real site scan from #5774 obfuscated)

image

### CODE REVIEW

**Review by valentijnscholten — APPROVED** [review]

File: dojo/models.py [hunk 1]
```diff
@@ -2166,17 +2166,14 @@ def duplicate_finding_set(self):
...
-        if scanner_confidence:
-            if scanner_confidence <= 2:
-                scanner_confidence_text = "Certain"
-            elif scanner_confidence >= 3 and scanner_confidence <= 5:
-                scanner_confidence_text = "Firm"
-            elif scanner_confidence >= 6:
-                scanner_confidence_text = "Tentative"
-
-        return scanner_confidence_text
+        if isinstance(self.scanner_confidence, int):
```
Comment (StefanFl) [inline dojo/models.py:2169]:
```suggestion
        if self.scanner_confidence and isinstance(self.scanner_confidence, int):
```

This would make the function more robust.

File: dojo/tools/wpscan/parser.py [hunk 2]
```diff
@@ -70,46 +71,65 @@ def get_findings(self, file, test):
...
-            self.get_vulnerabilities(report_date, node.get('vulnerabilities'), dupes, node, plugin)
+        for plugin in tree.get("plugins", []):
+            node = tree["plugins"][plugin]
+            self.get_vulnerabilities(
+                report_date,
+                node.get("vulnerabilities"),
+                dupes,
+                node,
+                plugin,
+                detection_confidence=tree["plugins"][plugin].get("confidence"),
```
Comment (StefanFl) [inline dojo/tools/wpscan/parser.py:87]:
```suggestion
                detection_confidence=node.get("confidence"),
```

This should work as well but would be easier to read.

**Comment by damiencarol** [comment]
made some changes for @StefanFl feedback

**Comment by StefanFl** [comment]
One of my 2 change requests makes a problem :-(

File: unittests/tools/test_wpscan_parser.py [hunk 3]
```diff
@@ -74,18 +74,24 @@ def test_parse_file_with_multiple_vuln(self):
...
-        self.assertEqual(datetime.datetime(2021, 3, 17, 12, 21, 6), finding.date)
+
+        with self.subTest(i=0):
+            finding = findings[0]
+            self.assertEqual("7391118e-eef5-4ff8-a8ea-f6b65f442c63", finding.unique_id_from_tool)
+            self.assertNotEqual("Info", finding.severity)  # it is a vulnerability so not 'Info'
+            self.assertEqual("Contact Form 7 < 5.3.2 - Unrestricted File Upload", finding.title)
+            self.assertEqual("CVE-2020-35489", finding.cve)
+            self.assertEqual(datetime.datetime(2021, 3, 17, 12, 21, 6), finding.date)
+            self.assertEqual("Certain", finding.get_scanner_confidence_text())  # data are => 100%
```
Comment (StefanFl) [inline unittests/tools/test_wpscan_parser.py:85]: This finding has no confidence, so the scanner confidence text should not be `Certain`

File: unittests/tools/test_wpscan_parser.py [hunk 4]
```diff
@@ -95,8 +101,52 @@ def test_parse_file_with_multiple_vuln_in_version(self):
...
-        self.assertEqual("WordPress 2.8.1-4.7.2 - Control Characters in Redirect URL Validation", finding.title)
-        self.assertEqual("fixed in : 4.6.4", finding.mitigation)
+
+        with self.subTest(i=2):
+            finding = findings[2]
+            self.assertEqual("d40374cf-ee95-40b7-9dd5-dbb160b877b1", finding.unique_id_from_tool)
+            self.assertNotEqual("Info", finding.severity)  # it is a vulnerability so not 'Info'
+            self.assertEqual("WordPress 2.8.1-4.7.2 - Control Characters in Redirect URL Validation", finding.title)
+            self.assertEqual("fixed in : 4.6.4", finding.mitigation)
+            self.assertEqual("Certain", finding.get_scanner_confidence_text())  # data are => 100%
```
Comment (StefanFl) [inline unittests/tools/test_wpscan_parser.py:111]: This finding has no confidence, so the scanner confidence text should not be `Certain`

File: unittests/tools/test_wpscan_parser.py [hunk 5]
```diff
@@ -74,18 +74,24 @@ def test_parse_file_with_multiple_vuln(self):
...
+
+        with self.subTest(i=4):
+            finding = findings[4]
+            self.assertIsNone(finding.unique_id_from_tool)  # interesting findings are not vlunerability
+            self.assertEqual("Info", finding.severity)  # it is not a vulnerability so severity should be 'Info'
+            self.assertEqual("Interesting finding: WordPress readme found: http://example/readme.html", finding.title)
+            self.assertEqual(datetime.datetime(2021, 3, 17, 12, 21, 6), finding.date)
+            # self.assertEqual("Certain", finding.scanner_confidence)  # data are => "confidence": 100,
+            self.assertEqual("Certain", finding.get_scanner_confidence_text())  # data are => "confidence": 100,
```
Comment (StefanFl) [inline unittests/tools/test_wpscan_parser.py:94]: I am not sure why this test case fails

**Review by StefanFl — CHANGES_REQUESTED** [review]

File: unittests/tools/test_wpscan_parser.py (hunk 5 above)
Comment (damiencarol) [reply unittests/tools/test_wpscan_parser.py]: fixed it

File: unittests/tools/test_wpscan_parser.py (hunk 4 above)
Comment (damiencarol) [reply unittests/tools/test_wpscan_parser.py]: fixed it

File: unittests/tools/test_wpscan_parser.py (hunk 3 above)
Comment (damiencarol) [reply unittests/tools/test_wpscan_parser.py]: fixed it

**Review by StefanFl — APPROVED** [review]
